### ISSUE TITLE
Unable to calculate script_data_hash with custom plutus witness

### ISSUE THREAD

**Issue by Karakatiza666** [Emurgo/cardano-serialization-lib#572]
I am creating a transaction that has  plutus datum and redeemer in WitnessSet after it is built. But I then append another plutus datum to built witness set (that is not required for validators) as I want to make its datum hash discoverable. Now, calc_script_data_hash obviously doesn't know about additional plutus datums that are not in the builder, and I am not able to get a correct hash via hash_script_data(). How should I go about this?

**Comment by Karakatiza666** [comment]
Code of what I am trying to do:
```      builder.set_script_data_hash(LCSL.ScriptDataHash.from_hex('0024cdcde8506abe9838047ec4167a910eef7f3c4615da4ebefc30bab21deef9'))
      const built = builder.build_tx()
      const finalWitnesses = mergeWitnesses([built.witness_set(), witnesses])
      const redeemers = finalWitnesses.redeemers() ?? LCSL.Redeemers.new()
      const datums = (d => d ? cslClone(LCSL.PlutusList, d) : undefined)(finalWitnesses.plutus_data())
      const plutusDataHash = LCSL.hash_script_data(redeemers, costmdls, datums)
      builder.set_script_data_hash(plutusDataHash)
```
I merge witnesses in the same order as I do for submitting transaction.

**Comment by lisicky** [comment]
Hi @Karakatiza666 ! What is the purpose of adding a datum after building tx? `script_data_hash` it is a hash of cost models, datums, and redeemers that are used in a tx, but if you calculate `script_data_hash` with a datum that is not used in a tx, that tx will fail when you try to submit it. Also, if you try to add datum to tx after `build_tx` the tx size will be increased, and after that, you will need to recalculate the tx fee. But a new fee value will change the tx body hash, and If you already signed the tx, that signature will be invalid.

**Comment by Karakatiza666** [comment]
Hey @lisicky ! I add datum to a custom TransactionWitnessSet before I calculate the size of and sign the transaction. When I am creating new script UTxO with new datum hash attached and do not add datum to witness set its value is not discoverable via datum hash through blockchain explorers. When datum is present in the witness set - explorers pick it up just fine.
Right now I am employing a workaround of parsing PPViewHashesDontMatch error message and re-signing the tx with the correct script_data_hash.

**Comment by lisicky** [comment]
@Karakatiza666 I am a little bit confused :D Why you don't just add all datums before `build_tx`? Could you describe your actions step by step?

**Comment by Karakatiza666** [comment]
I have found no way to add datum as a witness to the builder itself, not as a part of redeemer or spending script UTxO. So first I am calling build_tx() so that MintBuilder witnesses get processed and added to other builder redeemers. Then I add my custom datum witness to witnesses from built Transaction. Then I re-assemble transaction by reusing the Transaction body, aux_data, and all the witnesses together. I know that transaction's size and fee changes, it is accounted for when constructing the TX in the first place.

**Comment by lisicky** [comment]
Could you explain why you need put a datum without any connection with a tx input ? It might help to make a correct answer for you.
Usually workflow with CSL would be:
1. Add all outputs, mints, inputs and etc. to a builder
2. Call `builder. calc_script_data_hash()` if you have plutus scripts inputs or plutus script mints
3. Call `builder. add_change_if_needed()`
4. Call `builder. build_tx()` to get tx 
5. Sign tx and add signs to tx witness tx

All datums should be added during the first step. And you will probably get the error `NonOutputSupplimentaryDatums` from a cardano-node if will try to add a datum that is not connected with plutus script inputs.

**Comment by jeffhappily** [comment]
I have similar issue I think, what I'm trying to do is to add a datum for a reference input and I'm not sure how to do it since the `add_reference_input` function only takes in a UTXO. Does that mean I will have to manually set the witness?

**Comment by Karakatiza666** [comment]
@lisicky I mentioned before but here is a concrete example. Say I am creating UTxO at a script address, and attaching datum hash to it. Datum for this datum hash was never seen in Cardano before. After such transaction I am not able to get datum value from blockchain explorers like https://docs.blockfrost.io/#tag/Cardano-Scripts/paths/~1scripts~1datum~1%7Bdatum_hash%7D/get . But if I add a datum value for this hash to transaction witness set manually - while unrelated to any inputs and not used for validation, it allows blockchain explorers to have an association of datum hash to its value since its first appearance on the chain. In my (recent) experience I have submitted such transactions and this approach works as I described, and node does not return `NonOutputSupplimentaryDatums` error.

**Comment by Karakatiza666** [comment]
@jeffhappily in your case do you need to add datum to access its value in your validator logic? 
@lisicky also see https://cips.cardano.org/cips/cip31/ , Extra Datums paragraph

**Comment by jeffhappily** [comment]
Yeah, I mean there are other validators that need to read the datum value of this reference input.

**Comment by lisicky** [comment]
@Karakatiza666 thanks for the link. I need to check smth and if everything be ok we will add a new function "add_extra_datum".
@jeffhappily if you use reference input for a datum or script you don't need add that datum or script to a tx witnesses set

**Comment by jeffhappily** [comment]
@lisicky Thanks for the reference! But I don't think I get it 100%, for example, if I want to reference a UTXO (without spending it), I suppose I can use `txBuilder.add_reference_input`, but I assume this is when there's an inline datum in the UTXO, what happens if the UTXO that I want to use as a reference, only contains a datum hash without a datum, in this case I suppose I need to manually include the datum in the transaction if my other scripts need to read the datum of the UTXO?

Also, to clarify,
```rs
pub enum PlutusScriptSourceEnum {
    Script(PlutusScript),
    RefInput(TransactionInput, ScriptHash, Option<Language>),
}

pub enum DatumSourceEnum {
    Datum(PlutusData),
    RefInput(TransactionInput),
}
```

`PlutusScriptSourceEnum::Script` means normal script (by normal I mean not referenced where the script needs to be included in the transaction) whereas `PlutusScriptSourceEnum::RefInput` means reference script. `DatumSourceEnum::Datum` means datum hash whereas `DatumSourceEnum::RefInput` means inline datum, do I understand it correctly?

So `PlutusScriptSourceEnum::Script` + `DatumSourceEnum::RefInput` means a UTXO with a normal script _and_ an inline datum whereas  `PlutusScriptSourceEnum::RefInput` + `DatumSourceEnum::Datum` means a UTXO with a referenced script _and_ a datum hash, is that right?

**Comment by lisicky** [comment]
@jeffhappily to spend UTXO that locked by a Plutus script before Babbage, you were needed to add that UTXO as input and add Plutus script, datum, and redeemer to a witness set. Usually, it can be done by `add_plutus_script_input` function where you specify Plutus witnesses (script, datum, redeemer). After that tx builder put everything in the right places when you call `buld_tx`. 
But after babbage era arrived, you can use reference input instead of putting script or datum to a witness set. For that feature PlutusWitness struct was extended with a new function - `new_with_ref`. `new_with_ref` accepts `PlutusScriptSource` and `DatumSource` as arguments. It duplicates the previous logic if you use `PlutusScriptSource::new()` or `DatumSource::new()`, but if you call `new_ref_input()` instead of `new()` logic will be different, `new_ref_input()` accepts a tx input that will be used as a reference input, and after building tx that input will be in a reference inputs field in a tx body.
Long story short: utxo in ref inputs should be utxos with inlined script or datum, if you use ref input via `PlutusScriptSource:: new_ref_input` or `DatumSource:: new_ref_input` or `tx_builder.add_reference_input` you don't need to add that script or datum that referenced by a ref input to a witness set.

**Comment by lisicky** [comment]
> if I want to reference a UTXO (without spending it), I suppose I can use `txBuilder.add_reference_input`, but I assume this is when there's an inline datum in the UTXO, what happens if the UTXO that I want to use as a reference, only contains a datum hash without a datum, in this case I suppose I need to manually include the datum in the transaction if my other scripts need to read the datum of the UTXO?

I don't see any logic to use an utxo without inlined datum or script as ref input, because that utxo doesn't have any that can be referenced. 

> `PlutusScriptSourceEnum::Script` means normal script (by normal I mean not referenced where the script needs to be included in the transaction) whereas `PlutusScriptSourceEnum::RefInput` means reference script. `DatumSourceEnum::Datum` means datum hash whereas `DatumSourceEnum::RefInput` means inline datum, do I understand it correctly?

`PlutusScriptSourceEnum::Script` - just a script that will be placed into witness set
`PlutusScriptSourceEnum::RefInput` - ref input that will be placed into reference inputs filed
`DatumSourceEnum::Datum` - datum that  will be placed into witness set
`DatumSourceEnum::RefInput` - ref input that will be placed into reference inputs filed

`PlutusScriptSourceEnum::RefInput` + `DatumSourceEnum::Datum `- means ref input with script that you don't need to add to a witness set, because the script will be referenced and a datum the will be places to a witnesses set

**Comment by jeffhappily** [comment]
@lisicky So I'm trying to consume a UTxO with an inline datum, I suppose this is how I do it

```ts
const witness = CardanoWasm.PlutusWitness.new_with_ref(
  CardanoWasm.PlutusScriptSource.new(script),
  CardanoWasm.DatumSource.new_ref_input(myInput),
  redeemer
);
```
But one thing I notice is that in the transaction it generated, I see that this UTXO is _also_ being included in the referenced_inputs aside from the inputs, which is not what I'm expecting. I believe this is not the expected behavior but I could be wrong here, could you enlighten me on this?

**Comment by lisicky** [comment]
Hi @jeffhappily ! To use smth by reference (script or datum) you need to put a utxo that contains desired script or datum to the referenced_inputs field. That TransactionBuilder do if you use `DatumSource.new_ref_input` or `PlutusScriptSource.new_ref_input`

**Comment by jeffhappily** [comment]
@lisicky But for inline datum, I think the UTXO doesn't need to be included in the referenced_inputs field?

**Comment by lisicky** [comment]
@jeffhappily inlined datum you put into TransactionOutput

**Comment by jeffhappily** [comment]
@lisicky I mean to _consume_ a UTXO that contains an inline datum, you don't need to include it in your `referenced_input`, do you?

**Comment by lisicky** [comment]
@jeffhappily If you wanna just spend the utxo you need to it in inputs, but after that you can't use the utxo as reference input. But if you want to use an inline datum that contains the utxo, you need put the utxo into `referenced_inputs`
